fendermint/vm/interpreter/src/fvm/checkpoint.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use anyhow::{anyhow, Context};
+use ethers::types as et;
+use fendermint_vm_actor_interface::ipc::BottomUpCheckpoint;
+use fendermint_vm_genesis::{Power, Validator, ValidatorKey};
+use fendermint_vm_ipc_actors::gateway_router_facet::SubnetID;
+use fvm_ipld_blockstore::Blockstore;
+use tendermint::block::Height;
+use tendermint_rpc::{endpoint::validators, Client, Paging};
+
+use super::state::{ipc::GatewayCaller, FvmExecState};
+
+pub type Checkpoint = BottomUpCheckpoint;
+
+/// Validator voting power snapshot.
+#[derive(Debug, Clone)]
+pub struct PowerTable(pub Vec<Validator>);
+
+/// Changes in the power table.
+#[derive(Debug, Clone, Default)]
+pub struct PowerUpdates(pub Vec<Validator>);
+
+/// Construct and store a checkpoint if this is the end of the checkpoint period.
+/// Perform end-of-checkpoint-period transitions in the ledger.
+///
+/// If we are the boundary, return the validators eligible to sign and any updates
+/// to the power table, along with the checkpoint that needs to be signed by validators.
+pub async fn maybe_create_checkpoint<C, DB>(
+    client: &C,
+    gateway: &GatewayCaller<DB>,
+    state: &mut FvmExecState<DB>,
+) -> anyhow::Result<Option<(Checkpoint, PowerTable, PowerUpdates)>>
+where
+    C: Client + Sync + Send + 'static,
+    DB: Blockstore + Sync + Send + 'static,
+{
+    // Epoch transitions for checkpointing.
+    let height: tendermint::block::Height = state
+        .block_height()
+        .try_into()
+        .context("block height is not u64")?;
+
+    let block_hash = state
+        .block_hash()
+        .ok_or_else(|| anyhow!("block hash not set"))?;
+
+    match should_create_checkpoint(gateway, state, height)? {
+        None => Ok(None),
+        Some(subnet_id) => {
+            // Get the current power table.
+            let power_table = power_table(client, height)
+                .await
+                .context("failed to get the power table")?;
+
+            // TODO #252: Take the next changes from the gateway.
+            let power_updates = PowerUpdates(Vec::new());
```

Comment:
We assume that the updates have already been submitted and queued in the child's gateway through the top-down proof of finality, right?

Reply:
Yes, this takes the updates placed in the gateway by the top-down syncer.